lib/Onyx.js

```diff
@@ -551,34 +579,31 @@ function mergeCollection(collectionKey, collection) {
         }
     });
 
-    const existingKeyCollection = {};
-    const newCollection = {};
     return getAllKeys()
-        .then((keys) => {
-            _.each(collection, (data, dataKey) => {
-                if (keys.includes(dataKey)) {
-                    existingKeyCollection[dataKey] = data;
-                } else {
-                    newCollection[dataKey] = data;
-                }
-            });
-
-            const keyValuePairsForExistingCollection = _.reduce(existingKeyCollection, (finalArray, val, key) => ([
-                ...finalArray,
-                [key, JSON.stringify(val)],
-            ]), []);
-            const keyValuePairsForNewCollection = _.reduce(newCollection, (finalArray, val, key) => ([
-                ...finalArray,
-                [key, JSON.stringify(val)],
-            ]), []);
+        .then((persistedKeys) => {
+            // Split to keys that exist in storage and keys that don't
+            const [existingKeys, newKeys] = _.chain(collection)
+                .keys()
+                .partition(key => persistedKeys.includes(key))
+                .value();
+
+            const existingKeyCollection = _.pick(collection, existingKeys);
+            const newCollection = _.pick(collection, newKeys);
+            const keyValuePairsForExistingCollection = prepareKeyValuePairsForStorage(existingKeyCollection);
+            const keyValuePairsForNewCollection = prepareKeyValuePairsForStorage(newCollection);
 
             // New keys will be added via multiSet while existing keys will be updated using multiMerge
             // This is because setting a key that doesn't exist yet with multiMerge will throw errors
             const existingCollectionPromise = AsyncStorage.multiMerge(keyValuePairsForExistingCollection);
             const newCollectionPromise = AsyncStorage.multiSet(keyValuePairsForNewCollection);
 
+            // Capture (non-stringified) changes to cache
+            cache.merge(_.pairs(collection));
+
+            // Optimistically inform collection subscribers
+            keysChanged(collectionKey, collection);
```

Comment:
I think this maybe has the same problem I called out in another comment. The collection here is not referencing the cached value (which is a merged copy), but the value passed `mergeCollection()`?

Reply:
Alright, I appreciate all of the context here about references and I don't think we need to do anything yet with the `keysChanged()` stuff (my opinion). Although, you might want to clarify how things will work on the original issue to make sure the expectations are met with regards to the "reduced memory footprint" that some are expecting to see.

However... we must address how `mergeCollection()` works in this PR WRT to the cache merging because before the change we did:

- `AsyncStorage.multiMerge()` sets data in one way
- Subscribers are notified (with some receiving partial data atm)
- The data in storage matches what the next subscriber will get when they ask for the key

I will not recommend merging this PR if `cache.merge()` affects data differently than `AsyncStorage.multiMerge()`. The cache should reflect what is on disk... more on this in another comment.

Reply:
> However... we must address how `mergeCollection()` works in this PR WRT to the cache merging

Yep, I'll make sure cache reflects what's stored on disk for the merged collection

Reply:
`cache.merge(collection);` merges each collection item to cache and cache reflects what's stored on disk